### PR TITLE
Fix ipv6 vApp network creation

### PR DESCRIPTION
**Pull request by adezxc**
## Description
This PR fixes bug of not being able to create ipv6 vApp networks, as there was no `prefix_length` field, only `netmask`. With the addition of `prefix_length`, `netmask` is also marked as `Deprecated` as `prefix_length is more universal and can be used by both ipv4 and ipv6. 

Removes `netmask` field's `Default` (255.255.255.0) due to Terraform's limitations when using `ExactlyOneOf` or `ConflictsWith` together with `Default`. Because of this, users will be required to explicitly specify `netmask` in their `.tf` files or Terraform will re-create the resource.

All changes are related to `resourceVcdVappNetwork`

## Changes
* Add `prefix_length` to `resourceVcdVappNetwork`
* Add `ExactlyOneOf` to both `prefix_length` and `netmask` fields.
* Remove `Default` value of `netmask`
* Mark `netmask` as `Deprecated` as prefix length is supported by both ipv4 and ipv6.

### Testing
* Create `TestAccVcdVappNetwork_Isolated_ipv6` to test creation of ipv6 networks.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -60,3 +60,5 @@ require (
 	google.golang.org/grpc v1.51.0 // indirect
```
Comment (lvirbalas) [inline go.mod:60]: (Unrelated to this line). Please add the changelog entry ;)

File: website/docs/r/vapp_network.html.markdown [hunk 2]
```diff
@@ -50,7 +50,11 @@ The following arguments are supported:
 * `name` - (Required) A unique name for the network.
 * `description` - (Optional; *v2.7+*, *vCD 9.5+*) Description of vApp network
 * `vapp_name` - (Required) The vApp this network belongs to.
-* `netmask` - (Optional) The netmask for the new network. Default is `255.255.255.0`.
+* `netmask` - (Deprecated) Use `prefix_length` instead. The netmask for the new network.
```
Comment (lvirbalas) [inline website/docs/r/vapp_network.html.markdown:53]: My main concern is that we are removing the default value, which means users will have to amend existing configurations. This in turn means this is a breaking change, which would affect all configurations which have vApp network out there?

CC: @Didainius @dataclouder @adambarreiro

**Review by lvirbalas — CHANGES_REQUESTED** [review]

File: website/docs/r/vapp_network.html.markdown [hunk 3]
```diff
@@ -50,7 +50,11 @@ The following arguments are supported:
...
+
+~> **Note:** VCD returns Ipv6 addresses in extended-shortened format e.g `fe80:0:a:ab:0:abc:abcd:aaaa`, it is up to the user
```
Comment (Didainius) [inline website/docs/r/vapp_network.html.markdown:55]: I think this is the correct casing that people use for IPv6

```suggestion
~> **Note:** VCD returns IPv6 addresses in extended-shortened format e.g `fe80:0:a:ab:0:abc:abcd:aaaa`, it is up to the user
```

File: vcd/resource_vcd_vapp_network_test.go [hunk 4]
```diff
@@ -433,6 +632,46 @@ resource "vcd_vapp_network" "{{.resourceName}}" {
 }
 `
 
+const testAccCheckVappNetwork_update_ipv6 = `
```
Comment (Didainius) [inline vcd/resource_vcd_vapp_network_test.go:635]: Could you also add a `vcd_vapp_vm` that uses this IPv6 network to the test case? I don't think we have any VM using IPv6 because we haven't yet supported it, but we want to be sure there is no surprise in the VM itself while using this network with IPv6 address.

The cheapest one to add in terms of testing performance is "empty vm" (the one without vApp template so it saves time on cloning the image) https://registry.terraform.io/providers/vmware/vcd/latest/docs/resources/vapp_vm#example-usage-empty-vm

File: vcd/resource_vcd_vapp_network.go [hunk 5]
```diff
@@ -57,11 +57,22 @@ func resourceVcdVappNetwork() *schema.Resource {
...
-				Type:        schema.TypeString,
-				Optional:    true,
-				ForceNew:    true,
-				Default:     "255.255.255.0",
-				Description: "Netmask address for a subnet. Default is 255.255.255.0",
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				ForceNew:     true,
+				Deprecated:   "Use prefix_length instead",
```
Comment (Didainius) [inline vcd/resource_vcd_vapp_network.go:64]: Could be a good hint that `prefix_length` also supports IPv6
```suggestion
				Deprecated:   "Use prefix_length instead which supports both IPv4 and IPv6",
```

File: website/docs/r/vapp_network.html.markdown [hunk 6]
```diff
@@ -50,7 +50,11 @@ The following arguments are supported:
 * `name` - (Required) A unique name for the network.
 * `description` - (Optional; *v2.7+*, *vCD 9.5+*) Description of vApp network
 * `vapp_name` - (Required) The vApp this network belongs to.
-* `netmask` - (Optional) The netmask for the new network. Default is `255.255.255.0`.
+* `netmask` - (Deprecated) Use `prefix_length` instead. The netmask for the new network.
+
```
Comment (Didainius) [inline website/docs/r/vapp_network.html.markdown:54]: I can't comment in exact place, but as we are deprecating netmask, you should update the "## Example usage" snippet to use prefix_length. We avoid using deprecated attributes in our sample snippets.

Also, IMO, you could add another example snippet with IPv6 usage

**Review by Didainius — COMMENTED** [review]
First comments

**Comment by adezxc** [comment]
Putting this in draft as there is a bug related to updating vApp networks with DHCP.

File: website/docs/r/vapp_network.html.markdown [hunk 7]
```diff
@@ -57,8 +83,14 @@ The following arguments are supported:
 * `name` - (Required) A unique name for the network.
 * `description` - (Optional; *v2.7+*, *vCD 9.5+*) Description of vApp network
 * `vapp_name` - (Required) The vApp this network belongs to.
-* `netmask` - (Optional) The netmask for the new network. Default is `255.255.255.0`.
+* `netmask` - (Deprecated) Use `prefix_length` instead. The netmask for the new network.
+~> **Warning:** In `v3.9.0`, field `netmask` no longer has a `default` value of  `255.255.255.0` so that IPv6 can be supported using the new `prefix_length` field.
```
[the remark this answers is not in the document]
Comment (adezxc) [reply website/docs/r/vapp_network.html.markdown:88]: Thanks for the check. Fixed.

File: vcd/resource_vcd_vapp_network.go (hunk 5 above)
Comment (adezxc) [reply vcd/resource_vcd_vapp_network.go]: Fixed

File: website/docs/r/vapp_network.html.markdown (hunk 3 above)
Comment (adezxc) [reply website/docs/r/vapp_network.html.markdown]: Fixed

File: website/docs/r/vapp_network.html.markdown (hunk 6 above)
Comment (adezxc) [reply website/docs/r/vapp_network.html.markdown:87]: Fixed.

File: vcd/resource_vcd_vapp_network_test.go (hunk 4 above)
Comment (adezxc) [reply vcd/resource_vcd_vapp_network_test.go:636]: Added.

File: .changes/v3.9.0/1007-bug-fixes.md [hunk 8]
```diff
@@ -0,0 +1,3 @@
+* Added `prefix_length` field to `resourceVcdVappNetwork` to support creating IPv6 vApp networks. [GH-1007]
+* Deprecated `netmask` in `resourceVcdVappNetwork`. [GH-1007]
+* Removed `netmask` field's `Default` value which, if not provided before, will result in a Terraform error. The user would then need to add a `"netmask" = "255.255.255.0"` to their existing vApp networks. [GH-1007]
```
Comment (lvirbalas) [inline .changes/v3.9.0/1007-bug-fixes.md:3]: Nit: no need for dots at the end of changelog entries.

Comment (adezxc) [reply .changes/v3.9.0/1007-bug-fixes.md]: Fixed, muscle memory is still doing its thing

File: vcd/resource_vcd_vapp_network.go [hunk 9]
```diff
@@ -58,11 +58,22 @@ func resourceVcdVappNetwork() *schema.Resource {
...
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				ForceNew:     true,
+				Deprecated:   "Use prefix_length instead which supports both IPv4 and IPv6",
+				Description:  "Netmask address for a subnet.",
+				ExactlyOneOf: []string{"prefix_length"},
+			},
+			"prefix_length": {
+				Type:         schema.TypeString,
```
Comment (dataclouder) [inline vcd/resource_vcd_vapp_network.go:70]: This one has skipped my attention when I reviewed [PR 550](https://github.com/vmware/go-vcloud-director/pull/550/files). Why is this field a string and not an Int?
According to [the API](https://developer.vmware.com/apis/1507/vmware-cloud-director/doc/doc/types/IpScopeType.html), it should be an integer.

File: vcd/resource_vcd_vapp_network_test.go [hunk 10]
```diff
@@ -368,6 +511,62 @@ resource "vcd_vapp_network" "{{.resourceName}}" {
...
+resource "vcd_vapp_vm" "{{.vappVmName}}" {
+  org  = "{{.Org}}"
+  vdc  = "{{.Vdc}}"
+
+  vapp_name = "{{.vappName}}"
+  name      = "{{.vappVmName}}"
+  computer_name = "emptyVM"
+  memory        = 2048
+  cpus          = 2
+  cpu_cores     = 1
```
Comment (dataclouder) [inline vcd/resource_vcd_vapp_network_test.go:530]:
```suggestion
resource "vcd_vapp_vm" "{{.vappVmName}}" {
  org = "{{.Org}}"
  vdc = "{{.Vdc}}"

  vapp_name     = "{{.vappName}}"
  name          = "{{.vappVmName}}"
  computer_name = "emptyVM"
  memory        = 2048
  cpus          = 2
  cpu_cores     = 1
```

File: vcd/resource_vcd_vapp_network_test.go [hunk 11]
```diff
@@ -434,6 +633,62 @@ resource "vcd_vapp_network" "{{.resourceName}}" {
...
 `
 
+const testAccCheckVappNetwork_update_ipv6 = `
+resource "vcd_vapp" "{{.vappName}}" {
+  name = "{{.vappName}}"
+  org  = "{{.Org}}"
+  vdc  = "{{.Vdc}}"
+}
+
+resource "vcd_vapp_vm" "{{.vappVmName}}" {
```
Comment (dataclouder) [inline vcd/resource_vcd_vapp_network_test.go:643]: Script is not properly formatted

File: vcd/resource_vcd_vapp_network.go (hunk 9 above)
Comment (adezxc) [reply vcd/resource_vcd_vapp_network.go:70]: Mostly because `TypeInt` defaults to `0` and `TypeString` defaults to `""` and there could be an edge case where one creates a network with `prefix_length=0`, but also if someone provides a netmask, it seems wrong to have checks whether `netmask` is given, to then nullify the `prefix_length`. 

Took example from this code: https://github.com/vmware/terraform-provider-vcd/blob/main/vcd/resource_vcd_nsxt_alb_edgegateway_service_engine_group.go#L65-L73
cc @Didainius

File: vcd/resource_vcd_vapp_network_test.go (hunk 10 above)
Comment (adezxc) [reply vcd/resource_vcd_vapp_network_test.go]: Updated.

File: vcd/resource_vcd_vapp_network_test.go (hunk 11 above)
Comment (adezxc) [reply vcd/resource_vcd_vapp_network_test.go:643]: Fixed.

File: vcd/resource_vcd_vapp_network.go [hunk 12]
```diff
@@ -58,11 +58,22 @@ func resourceVcdVappNetwork() *schema.Resource {
...
-				ForceNew:    true,
-				Default:     "255.255.255.0",
-				Description: "Netmask address for a subnet. Default is 255.255.255.0",
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				ForceNew:     true,
+				Deprecated:   "Use prefix_length instead which supports both IPv4 and IPv6",
+				Description:  "Netmask address for a subnet.",
+				ExactlyOneOf: []string{"prefix_length"},
```
[the remark this answers is not in the document]
Comment (adezxc) [reply vcd/resource_vcd_vapp_network.go]: Fixed.

File: vcd/resource_vcd_vapp_network.go [hunk 13]
```diff
@@ -58,11 +58,22 @@ func resourceVcdVappNetwork() *schema.Resource {
...
+			},
+			"prefix_length": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				ForceNew:     true,
+				Description:  "Prefix length for a subnet",
+				ExactlyOneOf: []string{"netmask"},
```
[the remark this answers is not in the document]
Comment (adezxc) [reply vcd/resource_vcd_vapp_network.go]: Fixed.

File: .changes/v3.9.0/1007-bug-fixes.md [hunk 14]
```diff
@@ -0,0 +1,3 @@
+* Added `prefix_length` field to `resourceVcdVappNetwork` to support creating IPv6 vApp networks [GH-1007]
+* Deprecated `netmask` in `resourceVcdVappNetwork` [GH-1007]
+* Removed `netmask` field's `Default` value which, if not provided before, will result in a Terraform error. The user would then need to add a `"netmask" = "255.255.255.0"` to their existing vApp networks [GH-1007]
```
[the remark this answers is not in the document]
Comment (adezxc) [reply .changes/v3.9.0/1007-bug-fixes.md]: Updated. This confused me because in go-vcloud-director (https://github.com/vmware/go-vcloud-director/blob/main/CHANGELOG.md) everything is past tense, so didn't check how it's in this project.

File: website/docs/r/vapp_network.html.markdown [hunk 15]
```diff
@@ -15,14 +15,15 @@ Supported in provider *v2.1+*
 ## Example Usage
```
Comment (Didainius) [inline website/docs/r/vapp_network.html.markdown:15]:
```suggestion
## Example Usage (IPv4)
```

File: website/docs/r/vapp_network.html.markdown [hunk 16]
```diff
@@ -15,14 +15,15 @@ Supported in provider *v2.1+*
...
 
 ```hcl
+# IPv4 example
```
Comment (Didainius) [inline website/docs/r/vapp_network.html.markdown:18]: Our website docs usually have such structure (each example has its own H2 markdown tag (2 hashes '##').

As a result Terraform registry docs show quick links to each of the examples 
image


```suggestion
```hcl
```

File: website/docs/r/vapp_network.html.markdown [hunk 17]
```diff
@@ -45,6 +46,31 @@ resource "vcd_vapp_network" "vappNet" {
     end_address   = "192.168.2.50"
   }
 }
+
+# IPv6 example
```
Comment (Didainius) [inline website/docs/r/vapp_network.html.markdown:50]: As per the explanation above - this should be a "##" section of its own and then a new separate  hcl snippet  should be opened (I cannot put proper example suggestion because triple backticks "```" clash 


```suggestion
## Example usage (IPv6)
```

File: website/docs/r/vapp_network.html.markdown [hunk 18]
```diff
@@ -57,8 +92,15 @@ The following arguments are supported:
...
 * `vapp_name` - (Required) The vApp this network belongs to.
-* `netmask` - (Optional) The netmask for the new network. Default is `255.255.255.0`.
+* `netmask` - (Deprecated) Use `prefix_length` instead. The netmask for the new network.
+
+~> **Warning:** In `v3.9.0`, field `netmask` no longer has a `default` value of  `255.255.255.0` so that IPv6 can be supported using the new `prefix_length` field. 
+This change makes `terraform validate|plan` fail if the user didn't provide a value earlier and relied on default `255.255.255.0`.
+In case that happens, a user needs to add `"netmask" = "255.255.255.0"` to existing vApp networks.
+* `prefix_length` - (Optional) The subnet prefix length for the network.
 * `gateway` - (Required) The gateway for this network.
+~> **Note:** VCD returns IPv6 addresses in extended-shortened format e.g `fe80:0:a:ab:0:abc:abcd:aaaa`, it is up to the user
```
Comment (lvirbalas) [inline website/docs/r/vapp_network.html.markdown:102]: For some reason this note doesn't render:

Screenshot 2023-03-16 at 17 05 44

Comment (adezxc) [reply website/docs/r/vapp_network.html.markdown:103]: Fixed.

File: .changes/v3.9.0/1007-bug-fixes.md [hunk 19]
```diff
@@ -0,0 +1,3 @@
+* Add `prefix_length` field to `resourceVcdVappNetwork` to support creating IPv6 vApp networks [GH-1007]
+* Deprecate `netmask` in `resourceVcdVappNetwork` [GH-1007]
```
Comment (dataclouder) [inline .changes/v3.9.0/1007-bug-fixes.md:2]: This line should go in a file named `1007-deprecations.md`

File: .changes/v3.9.0/1007-bug-fixes.md [hunk 20]
```diff
@@ -0,0 +1,3 @@
+* Add `prefix_length` field to `resourceVcdVappNetwork` to support creating IPv6 vApp networks [GH-1007]
```
Comment (dataclouder) [inline .changes/v3.9.0/1007-bug-fixes.md:1]: If this is a bug fix, we need to state what the bug is, possibly with a reference to an issue report

Comment (adezxc) [reply .changes/v3.9.0/1007-bug-fixes.md:1]: Should I link the Github issue aswell, something in form of 
```
* Add `prefix_length` field to `resourceVcdVappNetwork` to support creating IPv6 vApp networks. This field is necessary as netmask representation is not suitable for IPv6  [GH-999] [GH-1007]
```

File: .changes/v3.9.0/1007-bug-fixes.md (hunk 19 above)
Comment (adezxc) [reply .changes/v3.9.0/1007-bug-fixes.md:2]: Updated.

File: .changes/v3.9.0/1007-bug-fixes.md (hunk 20 above)
Comment (lvirbalas) [reply .changes/v3.9.0/1007-bug-fixes.md:1]: Just noticed: `resourceVcdVappNetwork` - we should state user-facing resource name here.

File: .changes/v3.9.0/1007-bug-fixes.md (hunk 19 above)
Comment (lvirbalas) [reply .changes/v3.9.0/1007-bug-fixes.md:2]: Here too: `resourceVcdVappNetwork` - we should state user-facing resource name here.

File: .changes/v3.9.0/1007-bug-fixes.md [hunk 21]
```diff
@@ -0,0 +1,3 @@
...
+* Deprecate `netmask` in `resourceVcdVappNetwork` [GH-1007]
+* Remove `netmask` field's `Default` value which, if not provided before, will result in a Terraform error. The user would then need to add a `"netmask" = "255.255.255.0"` to their existing vApp networks [GH-1007]
```
Comment (lvirbalas) [inline .changes/v3.9.0/1007-bug-fixes.md:3]: `Remove `netmask` field's `Default` value` - please add from which resource too.

File: .changes/v3.9.0/1007-bug-fixes.md (hunk 20 above)
Comment (dataclouder) [reply .changes/v3.9.0/1007-bug-fixes.md:1]: The line should be something like this.
```
* Add `prefix_length` field to `resourceVcdVappNetwork` as creating IPv6 vApp networks was not supported due to the lack of a suitable subnet representation (Issue #999) [GH-1007]
```

However, if we are making change that will affect existing scripts because of a bug, we should also state that the default value we are removing was incorrect. This is a legitimate change in semantic versioning: we can't change what we have released, but we must change what has incorrect behavior, even if it breaks user workflows.

File: .changes/v3.9.0/1007-bug-fixes.md (hunk 21 above)
Comment (adezxc) [reply .changes/v3.9.0/1007-bug-fixes.md]: Fixed.

File: .changes/v3.9.0/1007-deprecations.md [hunk 22]
```diff
@@ -0,0 +1 @@
+* Deprecate `netmask` in `resourceVcdVappNetwork` [GH-1007]
```
[the remark this answers is not in the document]
Comment (adezxc) [reply .changes/v3.9.0/1007-deprecations.md]: Fixed.

File: .changes/v3.9.0/1007-bug-fixes.md (hunk 20 above)
Comment (adezxc) [reply .changes/v3.9.0/1007-bug-fixes.md]: Updated. I'm not sure if this is clear enough or if I maybe should be less explicit in a changelog entry

File: .changes/v3.9.0/1007-bug-fixes.md [hunk 23]
```diff
@@ -0,0 +1,2 @@
+* Add `prefix_length` field to `vcd_vapp_network` as creating IPv6 vApp networks was not supported due to the lack of a suitable subnet representation (Issue #999) [GH-1007]
+* Remove `vcd_vapp_network` `netmask` field's `Default` value as due to Terraform's limitation, it wouldn't allow for using only  `netmask` or `prefix_length` which is required for IPv6 networks. If no value was provided before, running the configuration will result in a Terraform error. The user would then need to add a `"netmask" = "255.255.255.0"` to their existing vApp networks [GH-1007]
```
[the remark this answers is not in the document]
Comment (adezxc) [reply .changes/v3.9.0/1007-bug-fixes.md]: Understood, I agree your version of the text is better, updated.

**Review by Didainius — APPROVED** [review]

**Review by lvirbalas — APPROVED** [review]

**Review by dataclouder — APPROVED** [review]
